### ISSUE TITLE
Regex collections should implement generic collection interfaces

### ISSUE THREAD

**Issue by justinvp** [dotnet/corefx#271]
`CaptureCollection`, `GroupCollection`, and `MatchCollection` currently only implement the non-generic `ICollection` interface. These collections should implement the generic collection interfaces to better interoperate with more modern APIs, such as LINQ. Since these collections are already indexable, they should implement  `IList<T>` and `IReadOnlyList<T>`, as well as the non-generic `IList` (to be consistent with the generic interfaces).

## Rationale and Usage

This is certainly a nice-to-have, but it *is* a long-standing [request](https://connect.microsoft.com/VisualStudio/feedback/details/469250/capturecollection-groupcollection-and-matchcollection-do-not-implement-icollection-t-and-ienumerable-t) that [developers](http://stackoverflow.com/q/7274422) [still](http://jimlynn.wordpress.com/2009/11/04/why-cant-i-use-linq-on-a-capturecollection/) [ask](http://stackoverflow.com/q/16379433) [about](http://stackoverflow.com/q/4108956). Implementing the generic interfaces will allow these collections to be used more easily with LINQ and interoperate better with more modern framework and library APIs.

For example, to use these collections with LINQ right now you have to know about and remember to use `Enumerable.Cast<TSource>()` to cast the non-generic `IEnumerable` into an `IEnumerable<T>`:

```c#
var captures =
    from capture in match.Groups.Cast<Group>().Last().Captures.Cast<Capture>()
    select capture.Value;
```

With these changes you'd no longer have to do that:

```c#
var captures =
    from capture in match.Groups.Last().Captures
    select capture.Value;
```

Plus, in the above example, you'd get a performance improvement when using `Enumerable.Last<TSource>.()` as its implementation has a fast-path for collections that implement `IList<T>`.

## Proposed API

```c#
// DebuggerDisplay and DebuggerTypeProxy added
[DebuggerDisplay("Count = {Count}")]
[DebuggerTypeProxy(typeof(RegexCollectionDebuggerProxy<Capture>))]
// Previously only implemented ICollection
public class CaptureCollection : IList<Capture>, IReadOnlyList<Capture>, IList
{
    // Existing members
    public int Count { get; }
    public Capture this[int i] { get; }
    public IEnumerator GetEnumerator();
    object ICollection.SyncRoot { get; }
    bool ICollection.IsSynchronized { get; }
    void ICollection.CopyTo(Array array, int arrayIndex);

    // Proposed members
    public void CopyTo(Capture[] array, int arrayIndex);
    IEnumerator<Capture> IEnumerable<Capture>.GetEnumerator();
    int IList<Capture>.IndexOf(Capture item);
    void IList<Capture>.Insert(int index, Capture item);
    void IList<Capture>.RemoveAt(int index);
    Capture IList<Capture>.this[int index] { get; set; }
    void ICollection<Capture>.Add(Capture item);
    void ICollection<Capture>.Clear();
    bool ICollection<Capture>.Contains(Capture item);
    bool ICollection<Capture>.IsReadOnly { get; }
    bool ICollection<Capture>.Remove(Capture item);
    int IList.Add(object value);
    void IList.Clear();
    bool IList.Contains(object value);
    int IList.IndexOf(object value);
    IList.Insert(int index, object value);
    bool IList.IsFixedSize { get; }
    bool IList.IsReadOnly { get; }
    void IList.Remove(object value);
    void IList.RemoveAt(int index);
    object IList.this[int index] { get; set; }
}

// DebuggerDisplay and DebuggerTypeProxy added
[DebuggerDisplay("Count = {Count}")]
[DebuggerTypeProxy(typeof(RegexCollectionDebuggerProxy<Group>))]
// Previously only implemented ICollection
public class GroupCollection : IList<Group>, IReadOnlyList<Group>, IList
{
    // Existing members
    public int Count { get; }
    public Group this[int groupnum] { get; }
    public Group this[String groupname] { get; }
    public IEnumerator GetEnumerator();
    object ICollection.SyncRoot { get; }
    bool ICollection.IsSynchronized { get; }
    void ICollection.CopyTo(Array array, int arrayIndex);

    // Proposed members
    public void CopyTo(Group[] array, int arrayIndex);
    IEnumerator<Group> IEnumerable<Group>.GetEnumerator();
    int IList<Group>.IndexOf(Group item);
    void IList<Group>.Insert(int index, Group item);
    void IList<Group>.RemoveAt(int index);
    Group IList<Group>.this[int index] { get; set; }
    void ICollection<Group>.Add(Group item);
    void ICollection<Group>.Clear();
    bool ICollection<Group>.Contains(Group item);
    bool ICollection<Group>.IsReadOnly { get; }
    bool ICollection<Group>.Remove(Group item);
    int IList.Add(object value);
    void IList.Clear();
    bool IList.Contains(object value);
    int IList.IndexOf(object value);
    IList.Insert(int index, object value);
    bool IList.IsFixedSize { get; }
    bool IList.IsReadOnly { get; }
    void IList.Remove(object value);
    void IList.RemoveAt(int index);
    object IList.this[int index] { get; set; }
}

// DebuggerDisplay and DebuggerTypeProxy added
[DebuggerDisplay("Count = {Count}")]
[DebuggerTypeProxy(typeof(RegexCollectionDebuggerProxy<Match>))]
// Previously only implemented ICollection
public class MatchCollection : IList<Match>, IReadOnlyList<Match>, IList
{
    // Existing members
    public int Count { get; }
    public virtual Match this[int i] { get; }
    public IEnumerator GetEnumerator();
    object ICollection.SyncRoot { get; }
    bool ICollection.IsSynchronized { get; }
    void ICollection.CopyTo(Array array, int arrayIndex);

    // Proposed members
    public void CopyTo(Match[] array, int arrayIndex);
    IEnumerator<Match> IEnumerable<Match>.GetEnumerator();
    int IList<Match>.IndexOf(Match item);
    void IList<Match>.Insert(int index, Match item);
    void IList<Match>.RemoveAt(int index);
    Match IList<Match>.this[int index] { get; set; }
    void ICollection<Match>.Add(Match item);
    void ICollection<Match>.Clear();
    bool ICollection<Match>.Contains(Match item);
    bool ICollection<Match>.IsReadOnly { get; }
    bool ICollection<Match>.Remove(Match item);
    int IList.Add(object value);
    void IList.Clear();
    bool IList.Contains(object value);
    int IList.IndexOf(object value);
    IList.Insert(int index, object value);
    bool IList.IsFixedSize { get; }
    bool IList.IsReadOnly { get; }
    void IList.Remove(object value);
    void IList.RemoveAt(int index);
    object IList.this[int index] { get; set; }
}

```

## Details

  * There was some discussion as to whether only the read-only interfaces should be implemented, or both the read-only and mutable interfaces. The consensus is to implement both the read-only and mutable interfaces. This is consistent with other collections in the framework. The mutable interfaces are implemented as read-only: mutable members are implemented explicitly and throw `NotSupportedException` (like `ReadOnlyCollection<T>`).

  * There was an open question as to whether the non-generic `IList` should be implemented as well. These collections are indexable and if  `IList<T>` and `IReadOnlyList<T>` are being implemented, `IList` should be implemented as well. This does add several more members, but they are all implemented explicitly so they don't add any new public members to intellisense, and the implementations are very straightforward.

  * `ICollection<T>.CopyTo` is implemented implicitly (public).

  * All other new members are implemented explicitly (non-public):

    * Mutable members are implemented explicitly because these collections are read-only and the mutable members throw `NotSupportedException` (like `ReadOnlyCollection<T>`).

    * `IList` members are implemented explicitly to hide non-generic members from intellisense.

    * `IList<T>.IndexOf` and `ICollection<T>.Contains` are implemented explicitly because these methods aren't very useful for these collections and should not be visible in intellisense by default. They're not useful because an implementation using `EqualityComparer<T>.Default` (consistent with other collections) will search the collection using reference equality due to the fact that `Capture`, `Group`, and `Match` do not implement `IEquatable<T>` and do not override `Equals()` and `GetHashCode()`. Further, these types do not have public constructors -- they are created internally by the regex engine, making it very unlikely that you'd want to search for an item in a collection "A" that was obtained from collection "B".

    * `IEnumerable<T>.GetEnumerator()` must be implemented explicitly because the non-generic `IEnumerable.GetEnumerator()` is already implemented implicitly and we can't overload on return type. This also precludes returning a `struct` `Enumerator` (for better enumeration performance) because changing the return type of the existing method would be a binary breaking change. As a result, you'll still have to specify the type when using `foreach` (e.g. `foreach (Capture capture in captures)`); you won't be able to use `var` (e.g. `foreach (var capture in captures)`), unfortunately.

## Open Questions

 * Should `GroupCollection` implement `IDictionary<string, Group>`, `IReadOnlyDictionary<string, Group>`, and `IDictionary`? `GroupCollection` already has a string indexer. Is it worth implementing the dictionary interfaces as part of this? Personally, I'm leaning toward "no" because there isn't a compelling scenario for the dictionary interfaces, and they can always be added in the future when needed.

## Pull Request

A PR with the proposed changes is available: #1756

## Updates

 * Edited this description to make it more of a *speclet*, based on the discussion below and the proposed API Review process.
 * Some improvements based on feedback from @sharwell.
 * Fixed existing members.
 * Added `IList`. These collections are indexable and it would be strange if `IList<T>` and `IReadOnlyList<T>` were implemented alongside `ICollection` but without `IList`.
 * Added `DebuggerDisplay` and `DebuggerTypeProxy` attributes.
 * Made `ICollection<T>.CopyTo` implicit (public).

**Comment by thomaslevesque** [comment]
> `IEnumerable<T>.GetEnumerator()` must be implemented explicitly because the non-generic `IEnumerable.GetEnumerator()` is already implemented implicitly and we can't overload on return type.

So it would break binary compatibility, right? But not source compatibility?

But is it possible for an assembly compiled against .NET 4.5 (or earlier) to run against .NET Core? If not, the argument of binary compatibility is moot, since the code will have to be recompiled anyway... Implementing the generic `GetEnumerator()` implicitly would be great, since it would make it possible to write `foreach (var capture in captures)`. OK, having to specify the type explicitly is not that bad, but I think it's confusing for beginners: why can I sometimes use `var`, and sometimes not?

**Comment by justinvp** [comment]
> But is it possible for an assembly compiled against .NET 4.5 (or earlier) to run against .NET Core? If not, the argument of binary compatibility is moot, since the code will have to be recompiled anyway... Implementing the generic GetEnumerator() implicitly would be great, since it would make it possible to write foreach (var capture in captures).

That is an excellent point, @thomaslevesque! Personally, I'd really like to be able to expose the generic `GetEnumerator()` to allow `var` in `foreach`. Since this is a new library in .NET Core, this may be the best opportunity to make such a change (before developers take a binary dependency on this library). It'd be a source compatible change, so any existing code ported to .NET Core would continue to work when recompiled, without the developer having to make any changes.

Further, if this change is possible, we may also want to take the opportunity to expose a `struct` enumerator to avoid the enumerator allocation.

I'd be happy to make such changes to the PR, if the owners agree the change can be made.

**Comment by davkean** [comment]
You need to o be able to build against this library and run on .NET Framework 4.6.

Sent from my Windows Phone
________________________________

**Comment by thomaslevesque** [comment]
> You need to o be able to build against this library and run on .NET Framework 4.6.

**Comment by thomaslevesque** [comment]
> You need to o be able to build against this library and run on .NET Framework 4.6.

Ah, I *knew* I was missing something ;)

I guess it wouldn't matter for reference types, but it would break for value types...

**Comment by justinvp** [comment]
> You need to o be able to build against this library and run on .NET Framework 4.6.

So if you take this suggestion (adding the generic interfaces to these types), the changes will have to be added to .NET 4.6 as well? How much time is left to make changes to .NET 4.6?

**Comment by thomaslevesque** [comment]
Actually my previous comment doesn't make sense... we're talking about specific collections, and there are no value types involved here. But then, couldn't it work in .NET 4.6 ? If the library is bound at compile time to a `GetEnumerator` method that returns a non-generic `IEnumerator`, but the actual method at runtime returns a generic `IEnumerator<Capture>`, it should work, right? Since `IEnumerator<Capture>` inherits `IEnumerator`...

**Comment by weshaggard** [comment]
Yes @justinvp any types that already exist on the full desktop framework will need to be updated there as well if they are updated in .NET Core. We have a high level goal of making code written for .NET work no matter what .NET platform you are targeting. As for how much time is left to add new public APIs to .NET 4.6 at this point we are not accepting any more unless they are critical.

With that said we fully expect .NET Core to evolve much more quickly then the full .NET framework that ships in Windows. We just have to be aware that any changes need to eventually make it there as well so while thinking about breaking changes you cannot just consider .NET Core you must consider the other platforms as well. If we don't do that then we risk diverging the platforms and making .NET development as a whole more difficult.

**Comment by justinvp** [comment]
Thanks, @weshaggard. I think it'd be worth updating one of the wiki pages with this information as I don't believe it's explicitly stated anywhere currently. I opened #392 to track that.

**Comment by justinvp** [comment]
@davkean & @weshaggard, this is getting off-topic, but one more related question about the full .NET framework: if I have some code that I compile against the .NET Core regex library, and I run my compiled assembly on the full .NET framework, does it use the regex types in the .NET Core regex library or the regex types in the full .NET framework at runtime?

**Comment by weshaggard** [comment]
@justinvp it would run the code that ships in the .NET Framework. The assembly identity System.Text.RegularExpressions ships in the box and is in the GAC so unless there were other changes to the identity then you wouldn't be running the code from this .NET Core library.

**Comment by justinvp** [comment]
From the [API Review](https://github.com/dotnet/apireviews/blob/master/2015-01-14-misc/README.md#316-implement-ilistt-ireadonlylistt-and-ilist-on-regex-collections):

> The implemented strongly typed, non-mutating methods should be implemented implicitly, i.e. publicly, not explicitly

These are the three methods:
 - `CopyTo`
 - `Contains`
 - `IndexOf`

Why? Is this just to match `ReadOnlyCollection<T>`?

I can see `CopyTo`, but I'm having a hard time coming up with how anyone would use `Contains` or `IndexOf` with these collections.

**Comment by terrajobst** [comment]
Yes, it's basically consistency with `ReadOnlyCollection<T>`. It also seems weird to go out of ones way to exclude these APIs from being directly callable.

**Comment by justinvp** [comment]
Can you show me sample code of how you would use `Contains` or `IndexOf`? How are you getting an instance of `Capture`, `Group`, or `Match` to pass into `Contains` or `IndexOf`?

**Comment by justinvp** [comment]
@davkean, in the API Review, you call out that `CopyTo`, `Contains`, and `IndexOf` should be implicit... I can get on board with `CopyTo`, but I'm not sure about `Contains` and `IndexOf`. I made these explicit for a reason, mentioned in the *speclet* above:

> IList<T>.IndexOf and ICollection<T>.Contains are implemented explicitly because these methods aren't very useful for these collections and should not be visible in intellisense by default. They're not useful because an implementation using EqualityComparer<T>.Default (consistent with other collections) will search the collection using reference equality due to the fact that Capture, Group, and Match do not implement IEquatable<T> and do not override Equals() and GetHashCode(). Further, these types do not have public constructors -- they are created internally by the regex engine, making it very unlikely that you'd want to search for an item in a collection "A" that was obtained from collection "B".

**Comment by justinvp** [comment]
@davkean ?

**Comment by davkean** [comment]
Sorry, for the late reply. I agreed with your assessment - I can think of contrived examples, but nothing solid.

**Comment by terrajobst** [comment]
This issue was [reviewed today](https://github.com/dotnet/apireviews/tree/master/2015-04-28-misc#271-regex-collections-should-implement-generic-collection-interfaces). It looks good as proposed.

**Comment by justinvp** [comment]
PR #1756 is ready for review (against the future branch).

Note: I just want to call out that the PR has `CopyTo` as implicit (public) and `Contains` and `IndexOf` as explicit (non-public). This is a slight deviation from what was discussed in the most recent API review.

I had originally proposed `CopyTo`, `Contains`, and `IndexOf` as all being explicit. The feedback from the first API review was to make these implicit to match `ReadOnlyCollection<T>`, but the reasoning for why these were proposed in the speclet as explicit was not discussed in the review.

 * `CopyTo` was proposed as explicit to match the non-generic `CopyTo` which was explicit.

 * `Contains` and `IndexOf` were proposed as explicit because these methods aren't very useful for these collections (see details in the speclet above).

I was actually on board with the feedback to make `CopyTo` implicit (my original reasoning for `CopyTo` was flawed), but still had some concerns about `Contains` and `IndexOf` that hadn't been discussed in the review. In the follow-up discussion in the comments (above) @davkean agreed with my assessment.

I was planning to make `CopyTo` implicit, but didn't update the speclet reflect this, so when this was discussed again in the most recent API review, you just discussed having everything as explicit as originally proposed, and agreed that this was OK.

I just want to call out that I've deviated slightly from this, making `CopyTo` implicit, but keeping `Contains` and `IndexOf` explicit. I've updated the speclet above to show `CopyTo` as being implicit to match the PR.

Let me know if you'd like me to change `CopyTo` back to being explicit as originally proposed.

**Comment by davkean** [comment]
@justinvp Nope no need to change anything, agree with what you've done here. Can't review the code, but someone on the team will be able to.

**Comment by stephentoub** [comment]
Fixed with #1756

**Comment by weitzhandler** [comment]
About time!
